v2/spiffetls/tlsconfig/hooks.go

```diff
@@ -0,0 +1,12 @@
+package tlsconfig
+
+import (
+	"time"
+
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+)
+
+type Hooks struct {
+	GetTLSCertificateStart func(spiffeid.ID, time.Time)
```

Comment:
I think this API should work similarly to how it's done in [`httptrace`](https://golang.org/pkg/net/http/httptrace/#ClientTrace):

```go
type GetTLSCertificate struct{
   // details
}

type GotTLSCertificate struct{
   Start time.Time
   End time.Time
   Cert *tls.Certificate
   Err error
}

type ConfigTrace struct {
    GetTLSCertificate func(GetTLSCertificateInfo)
    GotTLSCertificate func(GotTLSCertificateInfo)
}
```

and then this func would become:
```go
func getTLSCertificate(svid x509svid.Source, hooks Hooks) (*tls.Certificate, error) {
	start := time.Now()
	if hooks.GetTLSCertificate != nil {
		hooks.GetTLSCertificate(GetTLSCertificate{Time: start})
	}

	s, err := svid.GetX509SVID()
	if err != nil {
		hooks.GotTLSCertificate(GotTLSCertificate{Start: start, End: time.Now(), Err: err})
		return nil, err
	}

	cert := &tls.Certificate{
		Certificate: make([][]byte, 0, len(s.Certificates)),
		PrivateKey:  s.PrivateKey,
	}

	for _, svidCert := range s.Certificates {
		cert.Certificate = append(cert.Certificate, svidCert.Raw)
	}

	if hooks.GotTLSCertificate != nil {
		hooks.GotTLSCertificate(GotTLSCertificate{Start: start, End: time.Now(), Cert: cert})
	}

	return cert, nil
}
```

Reply:
@aybabtme Thanks! My intention was to create `GetTLSCertificateStart` and `GetTLSCertificateEnd` which would be analogous to `ConnectStart` and `ConnectDone` in httptrace. I think we can achieve the same thing with start/end and is more obvious than get vs got.

```
func getTLSCertificate(svid x509svid.Source, hooks Hooks) (*tls.Certificate, error) {
	start := time.Now()
	if hooks.GetTLSCertificate != nil {
		hooks.GetTLSCertificateStart(GetTLSCertificateStart{Time: start})
	}

	s, err := svid.GetX509SVID()
	if err != nil {
		hooks.GetTLSCertificateEnd(GetTLSCertificateEnd{Start: start, End: time.Now(), Err: err})
		return nil, err
	}

	cert := &tls.Certificate{
		Certificate: make([][]byte, 0, len(s.Certificates)),
		PrivateKey:  s.PrivateKey,
	}

	for _, svidCert := range s.Certificates {
		cert.Certificate = append(cert.Certificate, svidCert.Raw)
	}

	if hooks.GotTLSCertificate != nil {
		hooks.GetTLSCertificateEnd(GetTLSCertificateEnd{Start: start, End: time.Now(), Cert: cert})
	}

	return cert, nil
}
```

Reply:
sgtm